homeassistant/components/goodwe/__init__.py

```diff
@@ -0,0 +1,121 @@
+"""The Goodwe inverter component."""
+from datetime import timedelta
+import logging
+
+from goodwe import InverterError, RequestFailedException, connect
+
+from homeassistant import config_entries, core
+from homeassistant.const import CONF_HOST, CONF_SCAN_INTERVAL
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    CONF_MODEL_FAMILY,
+    CONF_NETWORK_RETRIES,
+    CONF_NETWORK_TIMEOUT,
+    DEFAULT_NETWORK_RETRIES,
+    DEFAULT_NETWORK_TIMEOUT,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    KEY_COORDINATOR,
+    KEY_INVERTER,
+    PLATFORMS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: core.HomeAssistant, entry: config_entries.ConfigEntry
+):
+    """Set up the Goodwe components from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    name = entry.title
+    host = entry.data[CONF_HOST]
+    model_family = entry.data[CONF_MODEL_FAMILY]
+    scan_interval = entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+    network_retries = entry.options.get(CONF_NETWORK_RETRIES, DEFAULT_NETWORK_RETRIES)
+    network_timeout = entry.options.get(CONF_NETWORK_TIMEOUT, DEFAULT_NETWORK_TIMEOUT)
+
+    # Connect to Goodwe inverter
+    try:
+        inverter = await connect(
+            host=host,
+            family=model_family,
+            comm_addr=0,
+            timeout=network_timeout,
+            retries=network_retries,
+        )
+    except InverterError as err:
+        raise ConfigEntryNotReady from err
+
+    async def async_update_data():
+        """Fetch data from the inverter."""
+        try:
+            data = await inverter.read_runtime_data()
+        except RequestFailedException as ex:
+            # UDP communication with inverter is by definition unreliable.
+            # It is rather normal in many environments to fail to receive
+            # proper response in usual time, so we intentionally report
+            # failures only after consecutive streak of 3 of them.
+            if ex.consecutive_failures_count < 3:
+                # return empty dictionary
+                # sensors will keep their previous values
+                _LOGGER.debug(f"Request failed (#{ex.consecutive_failures_count}).")
+                return {}
+            else:
+                # return None
+                # sensors will report themselves as not available
+                _LOGGER.debug(
```

Comment:
Yes, this may sound weird at first look but the behavior is intentional and fine-tuned over last year from experience of many users. The 3 (silent) retries was selected as best approach for most users.
(There are users out there with various usage scenarios, some have 30s poll time just to watch it, some use aggressive 1s scan interval to use it for real-time control of devices consuming the produced power. To accommodate these different needs, this a bit unusual approach was selected).

When the inverter is not responding at all, the exception is intentionally not thrown, since it is usually a correct state of a solar inverter (non-battery) which went to sleep mode overnight.
All it's sensors will report themselves as Unavailable and will come back to life again in the morning when sun will wake up the inverter.